### ISSUE TITLE
Typescript 3.3.1 wrongly considers object as 'Possibly 'undefined''

### ISSUE THREAD

**Issue by vladeck** [Microsoft/TypeScript#29805]
**TypeScript Version:**  3.3.1

**Issue:** Even if the code is explicitly checking against variable not being `undefined` the compiler is complaining about `Object is possibly 'undefined'. ts(2532)` Strict checking is turned on with `"strict": true`.

![ts_issue_001](https://user-images.githubusercontent.com/16443718/52427958-4ee36d80-2b01-11e9-9618-97d78778dd9d.jpg)

![ts_issue_002](https://user-images.githubusercontent.com/16443718/52427967-5276f480-2b01-11e9-9758-c97a588b59c1.jpg)

**Comment by j-oliveras** [comment]
Duplicated/related to #9998.

Basically, typescript don't know if the function is called synchronously or not.

**Comment by vladeck** [comment]
Are you talking about `find` -> `predicate` function or the outter function of this code? Could you explain why this code compiles:

```js
if (objectToMoveParentPathComponent !== undefined) {

    const c = objectToMoveParentPathComponent;

    const objectToMoveParent = objectToMoveResult.parents.find(
        p => p.parentObjectId === c.objectId // No error
    );

    // ...

}
```

**Comment by j-oliveras** [comment]
The predicate.

This code compiles because you created a local const variable `c` that is know to be defined. This is the workaround to this issue.
